### PR TITLE
[codex] Implement DOM positioning helpers in ars-dom

### LINKED ISSUE #113: task: Implement containing-block detection and coordinate-space conversion in ars-dom positioning
**Wave 2-6c** | **Points: 5** | **Epic: #6 (DOM utilities)**

## Goal

Implement containing-block detection and coordinate-space conversion for DOM-positioned overlays so browser measurements can be translated into the correct local coordinate system for `Strategy::Absolute` and transformed ancestors.

## Layer

Subsystem

## Framework

None

## Test tier

Mixed

## Depends on

- #67 (pure geometry positioning algorithm)

## Spec refs

- `spec/foundation/11-dom-utilities.md` §2.3.1 "Step 0: Detect Containing Block" (L539)
- `spec/foundation/11-dom-utilities.md` §2.8 "Coordinate System" (L1045)
- `spec/foundation/11-dom-utilities.md` §2.8.1 "CSS Transform Ancestor Detection" (L1074)
- `spec/foundation/11-dom-utilities.md` §6.6 "CSS Containment Interaction" (L2434)

## Files to create/modify

- `crates/ars-dom/src/positioning/dom.rs` (new)
- `crates/ars-dom/src/positioning/mod.rs`
- `crates/ars-dom/src/lib.rs`
- `crates/ars-dom/Cargo.toml` (if additional DOM/CSS `web-sys` features are required)

## Tests to add first

- Unit tests for a pure helper that recognizes all spec-listed containing-block-causing CSS properties.
- Unit tests for client-space to local-space coordinate conversion against synthetic rects.
- Web-targeted smoke tests for `find_containing_block_ancestor()` returning the ancestor rect when transform/filter/contain/container-type properties are present.
- Unit tests for offset-parent adjustment used by `Strategy::Absolute`.

## Acceptance criteria

- `find_containing_block_ancestor(element) -> Option<Rect>` implemented for web targets.
- Detection covers `transform`, `perspective`, `filter`, `backdrop-filter`, `contain`, `content-visibility`, `container-type`, and relevant `will-change` values.
- Coordinate adjustment helpers convert viewport-relative measurements into the containing block or offset-parent local coordinate space.
- `Strategy::Absolute` conversion path is implemented as a reusable helper rather than adapter-local logic.
- Debug-mode warning helper exists for portal targets whose ancestors create containing blocks.
- Host/SSR builds remain safe via non-web stubs.

## Spec impact

No spec change required.

### PR DESCRIPTION
**Pull request by fogodev**
## Summary
- add DOM-backed positioning helpers in `ars-dom` for client-to-local coordinate conversion, containing-block lookup, offset-parent rect lookup, and portal/containment warnings
- keep CSS trigger classification private while re-exporting the new public helper surface through `ars_dom::positioning` and the crate root
- expand unit, host-web, and wasm browser coverage for containing-block detection and warning/no-op paths

## Why
- `#113` requires reusable DOM-side measurement and coordinate-space helpers around the pure `compute_position()` engine
- overlays need consistent containing-block and offset-parent detection before the shared glue-layer task wires these helpers into a full measurement pipeline

## Validation
- `cargo test -p ars-dom`
- `cargo xtest dom-browser`
- `cargo xci`

Closes #113

### CODE REVIEW

File: crates/ars-dom/src/positioning/dom.rs [hunk 1]
```diff
@@ -0,0 +1,995 @@
...
+#[cfg(all(feature = "web", not(target_arch = "wasm32")))]
+fn find_containing_block_ancestor_impl(_element: &web_sys::Element) -> Option<Rect> {
+    None
+}
+
+#[cfg(all(feature = "web", target_arch = "wasm32"))]
+fn offset_parent_rect_impl(element: &web_sys::Element) -> Option<Rect> {
+    let element = element.dyn_ref::<web_sys::HtmlElement>()?;
+    let offset_parent = element.offset_parent()?;
+    Some(dom_rect_to_rect(&offset_parent.get_bounding_client_rect()))
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-dom/src/positioning/dom.rs]: Fixed in `0e97397`.

`offset_parent_rect()` now returns the offset parent's padding-box rect instead of its border-box rect by adding `client_left()` / `client_top()` to the origin and using `client_width()` / `client_height()` for the size. That keeps `client_*_to_local_space()` aligned with how CSS resolves absolute `top` / `left`.

I also added a pure regression test for the padding-box conversion path, a wasm browser regression with non-zero borders, and updated `spec/foundation/11-dom-utilities.md` so the spec matches the browser behavior.

Validation:
- `cargo test -p ars-dom`
- `cargo xtest dom-browser`
- `cargo xci`

**Comment by fogodev** [comment]
@codex

File: crates/ars-dom/src/positioning/dom.rs [hunk 2]
```diff
@@ -0,0 +1,1057 @@
...
+
+#[cfg(all(feature = "web", target_arch = "wasm32"))]
+fn offset_parent_rect_impl(element: &web_sys::Element) -> Option<Rect> {
+    let element = element.dyn_ref::<web_sys::HtmlElement>()?;
+    let offset_parent = element.offset_parent()?;
+    let rect = offset_parent.get_bounding_client_rect();
+
+    Some(Rect {
+        x: rect.x() + f64::from(offset_parent.client_left()),
+        y: rect.y() + f64::from(offset_parent.client_top()),
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-dom/src/positioning/dom.rs]: Fixed in `2ef06c5`.

`offset_parent_rect()` now converts the offset parent into scroll-adjusted padding-box coordinates by subtracting `scroll_left()` / `scroll_top()` after computing the padding-box origin. I added both a pure regression test and a wasm browser regression that exercise a scrolled offset parent so `client_point_to_local_space()` stays aligned with CSS absolute positioning.

Validation:
- `cargo test -p ars-dom`
- `cargo xtest dom-browser`
- `cargo xci`

File: crates/ars-dom/src/positioning/dom.rs [hunk 3]
```diff
@@ -0,0 +1,1057 @@
...
+            current = ancestor.parent_element();
+
+            continue;
+        };
+
+        let snapshot = style_snapshot(&style);
+
+        if let Some(reason) = containing_block_reason(&snapshot) {
+            return Some((
+                dom_rect_to_rect(&ancestor.get_bounding_client_rect()),
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-dom/src/positioning/dom.rs]: Fixed in `2ef06c5`.

`walk_containing_block_ancestors()` now returns the containing block's padding-box rect instead of the raw border-box rect by reusing the shared padding-box helper. I added a wasm regression with bordered transformed ancestors and updated `spec/foundation/11-dom-utilities.md` so the documented contract matches the implementation.

Validation:
- `cargo test -p ars-dom`
- `cargo xtest dom-browser`
- `cargo xci`

File: crates/ars-dom/src/positioning/dom.rs [hunk 4]
```diff
@@ -0,0 +1,1135 @@
...
+#[cfg(all(feature = "web", not(target_arch = "wasm32")))]
+fn warn_if_floating_element_has_containment_issue_impl(_floating: &web_sys::Element) {}
+
+#[cfg(all(feature = "web", target_arch = "wasm32"))]
+fn walk_containing_block_ancestors(
+    element: &web_sys::Element,
+) -> Option<(Rect, ContainingBlockReason)> {
+    let window = web_sys::window()?;
+
+    let mut current = element.parent_element();
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-dom/src/positioning/dom.rs]: Fixed in `5f2985d`.

The containing-block and overlay-containment walks now follow the composed tree instead of stopping at `parent_element()`: they climb with `parent_node()` and, when they hit a `ShadowRoot`, continue from `ShadowRoot.host()`. That keeps transformed or contained shadow hosts in the local-space origin calculation.

I also added a wasm browser regression that mounts the floating element inside a shadow root under a transformed, bordered host, plus the matching spec update in `spec/foundation/11-dom-utilities.md`.

Validation:
- `cargo test -p ars-dom`
- `cargo xtest dom-browser`
- `cargo xci`

File: crates/ars-dom/src/positioning/dom.rs [hunk 5]
```diff
@@ -0,0 +1,1178 @@
...
+
+#[cfg(all(feature = "web", target_arch = "wasm32"))]
+use wasm_bindgen::JsCast;
+
+use super::Rect;
+
+/// Convert a client-space point into the local coordinate space of `local_origin`.
+#[must_use]
+pub fn client_point_to_local_space(x: f64, y: f64, local_origin: &Rect) -> (f64, f64) {
+    (x - local_origin.x, y - local_origin.y)
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-dom/src/positioning/dom.rs:21]: Fixed in `50c4640`.

The transformed-containing-block path is now explicitly guarded to axis-aligned translation-only transforms. For `transform`-established containing blocks, `find_containing_block_ancestor()` inspects the computed transform with `DomMatrixReadOnly`; if the matrix includes scale, rotation, skew, or other non-translation terms, the helper returns `None` instead of fabricating a wrong local-space origin.

I added a wasm regression with `scale(2)` that confirms the helper refuses the non-invertible rect-only conversion path, and updated `spec/foundation/11-dom-utilities.md` to document that matrix-aware conversion is required for non-translation transforms.

Validation:
- `cargo test -p ars-dom`
- `cargo xtest dom-browser`
- `cargo xci`

File: crates/ars-dom/src/positioning/dom.rs [hunk 6]
```diff
@@ -0,0 +1,1178 @@
...
+
+    None
+}
+
+#[cfg(all(feature = "web", target_arch = "wasm32"))]
+fn next_composed_ancestor_element(node: &web_sys::Node) -> Option<web_sys::Element> {
+    let parent = node.parent_node()?;
+
+    if let Some(element) = parent.dyn_ref::<web_sys::Element>() {
+        return Some(element.clone());
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-dom/src/positioning/dom.rs:407]: Fixed in `50c4640`.

The composed-ancestor walk now follows slot assignment before falling back to `parent_node()` / `ShadowRoot.host()`: if the current node is an `Element` with `assigned_slot()`, the walk continues from that slot so slotted light-DOM nodes can still discover transformed or contained ancestors inside the shadow tree.

I added a wasm regression that mounts a slotted light-DOM child under a transformed, bordered ancestor in shadow DOM and verifies `find_containing_block_ancestor()` returns that containing block, plus the matching spec update in `spec/foundation/11-dom-utilities.md`.

Validation:
- `cargo test -p ars-dom`
- `cargo xtest dom-browser`
- `cargo xci`

File: crates/ars-dom/src/positioning/dom.rs [hunk 7]
```diff
@@ -0,0 +1,1288 @@
...
+    if normalized.is_empty() || normalized == "none" {
+        return true;
+    }
+
+    let Ok(matrix) = web_sys::DomMatrixReadOnly::new_with_str(transform) else {
+        return false;
+    };
+
+    matrix.is_2d()
+        && (matrix.a() - 1.0).abs() < f64::EPSILON
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-dom/src/positioning/dom.rs]: Fixed in `7534907`.

`transform_is_translation_only()` now inspects the full `DomMatrixReadOnly` 4x4 matrix instead of requiring `is_2d()`. That keeps the guard strict for non-axis-aligned transforms, but it now accepts identity-plus-translation `matrix3d(...)` cases such as `translate3d(...)` and `translateZ(0)` so transformed containing blocks still participate in local-space conversion.

I added a wasm regression for a `translate3d(12px, 18px, 0)` containing block and updated `spec/foundation/11-dom-utilities.md` to document that translation-only `matrix3d` transforms are allowed while scaled, rotated, skewed, and perspective transforms still require matrix-aware conversion.

Validation:
- `cargo test -p ars-dom`
- `cargo xtest dom-browser`
- `cargo xci`

File: crates/ars-dom/src/positioning/dom.rs [hunk 8]
```diff
@@ -0,0 +1,1317 @@
...
+fn find_containing_block_ancestor_impl(_element: &web_sys::Element) -> Option<Rect> {
+    None
+}
+
+#[cfg(all(feature = "web", target_arch = "wasm32"))]
+fn offset_parent_rect_impl(element: &web_sys::Element) -> Option<Rect> {
+    let element = element.dyn_ref::<web_sys::HtmlElement>()?;
+    let offset_parent = element.offset_parent()?;
+    let mut rect = padding_box_rect(&offset_parent);
+    rect.x -= f64::from(offset_parent.scroll_left());
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-dom/src/positioning/dom.rs:247]: Fixed in `3aec935`.

`offset_parent_rect()` now applies the same safety rule as the containing-block path: after resolving the offset parent, it reads the computed `transform` and returns `None` unless the transform is translation-only. That prevents absolute-positioned overlays from using plain origin subtraction against scaled or rotated offset parents where `left` / `top` live in pre-transform local space.

I added wasm regressions for both sides of that contract: a `scale(2)` offset parent now returns `None`, while a translation-only `translate3d(12px, 18px, 0)` offset parent still returns the scroll-adjusted padding-box rect. The absolute-positioning section in `spec/foundation/11-dom-utilities.md` now documents the same `None` fallback for non-translation transforms.

Validation:
- `cargo test -p ars-dom`
- `cargo xtest dom-browser`
- `cargo xci`

File: crates/ars-dom/src/positioning/dom.rs [hunk 9]
```diff
@@ -0,0 +1,1317 @@
...
+
+    let Ok(matrix) = web_sys::DomMatrixReadOnly::new_with_str(transform) else {
+        return false;
+    };
+
+    is_identity_or_translation_component(matrix.m11(), 1.0)
+        && is_identity_or_translation_component(matrix.m12(), 0.0)
+        && is_identity_or_translation_component(matrix.m13(), 0.0)
+        && is_identity_or_translation_component(matrix.m14(), 0.0)
+        && is_identity_or_translation_component(matrix.m21(), 0.0)
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-dom/src/positioning/dom.rs:379]: Fixed in `3aec935`.

`transform_is_translation_only()` now explicitly requires `m43 == 0`, so only identity-plus-translation matrices in x/y remain eligible for rect-only local conversion. `translate3d(x, y, 0)` and `translateZ(0)` still pass, but any non-zero z translation now returns `None` instead of pretending the transform is axis-aligned.

I added a wasm regression for `transform: translateZ(8px)` to lock that down and updated `spec/foundation/11-dom-utilities.md` to state that non-zero z translation requires matrix-aware conversion just like scale, rotation, skew, and perspective.

Validation:
- `cargo test -p ars-dom`
- `cargo xtest dom-browser`
- `cargo xci`
